storage: throw if timequerying a closed log

andrwng opened the pull request:
This is follow-up to
https://github.com/redpanda-data/redpanda/pull/28328.

The previous PR added throws to timequery reader calls, but omitted the check for timequery() itself.

```
ERROR 2025-11-03 16:39:15,936 [shard 0:tran] assert - Assert failure: (src/v/storage/disk_log_impl.cc:3216) '!_closed' timequery on closed log - {offsets: {start_offset:0, committed_offset:0, committed_offset_term:1, dirty_offset:0, dirty_offset_term:1}, is_closed: true, segments: [{size: 1, [{offset_tracker:{term:1, base_offset:0, committed_offset:0, stable_offset:0, dirty_offset:0}, compacted_segment=0, finished_self_compaction=0, finished_windowed_compaction=0, generation=3, reader={/var/lib/redpanda/data/kafka/topic-fadnsotldv/6_28/0-1-v1.log, (155 bytes)}, writer={closed:true, fallocation_offset:155, stable_offset:155, flushed_offset:155, committed_offset:155, inflight: 0, bytes_flush_pending:0, stats: { merged_writes: 0, bytes_requested : 155, bytes_written : 4096, appends : 2, flushes: 2, fsyncs: 3, truncates: 1, fallocations: 1, last_page_hydrations: 0}}, cache={cache_size=0, dirty tracker: {min: -9223372036854775808, max: -9223372036854775808}}, compaction_index:nullopt, closed=1, tombstone=0, index={file:/var/lib/redpanda/data/kafka/topic-fadnsotldv/6_28/0-1-v1.base_index, offsets:0, index:{header_bitflags:0, base_offset:0, max_offset:0, base_timestamp:{timestamp: 1762187916005}, max_timestamp:{timestamp: 1762187916005}, batch_timestamps_are_monotonic:1, with_offset:true, non_data_timestamps:1, broker_timestamp:{nullopt}, num_compactible_records_appended:{0}, clean_compact_timestamp:{nullopt}, may_have_tombstone_records:1, self_compact_timestamp:{nullopt}, may_have_transaction_batches:1, index(1, 1, 1)}, step:32768, needs_persistence:0}}]}], config: {ntp: {kafka/topic-fadnsotldv/6}, base_dir: /var/lib/redpanda/data, overrides: {compaction_strategy: {nullopt}, cleanup_policy_bitflags: {delete}, segment_size: {nullopt}, retention_bytes: {{nullopt}}, retention_time_ms: {{nullopt}}, recovery_enabled: false, retention_local_target_bytes: {{nullopt}}, retention_local_target_ms: {{nullopt}}, remote_delete: {true}, segment_ms: {{nullopt}}, initial_retention_local_target_bytes: {{nullopt}}, initial_retention_local_target_ms: {{nullopt}}, write_caching: {nullopt}, flush_ms: {nullopt}, flush_bytes: {nullopt}, iceberg_mode: disabled, remote_allow_gaps: {nullopt}, delete_retention_ms: {disabled}, min_cleanable_dirty_ratio: {{nullopt}}, min_compaction_lag_ms: {nullopt}, max_compaction_lag_ms: {nullopt} }}, revision: 28, topic_revision: 28, remote_revision: 28}}
```

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes

* None
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
storage: throw if timequerying a closed log

This is follow-up to
https://github.com/redpanda-data/redpanda/pull/28328.

The previous PR added throws to timequery reader calls, but omitted the
check for timequery() itself.

```
ERROR 2025-11-03 16:39:15,936 [shard 0:tran] assert - Assert failure: (src/v/storage/disk_log_impl.cc:3216) '!_closed' timequery on closed log - {offsets: {start_offset:0, committed_offset:0, committed_offset_term:1, dirty_offset:0, dirty_offset_term:1}, is_closed: true, segments: [{size: 1, [{offset_tracker:{term:1, base_offset:0, committed_offset:0, stable_offset:0, dirty_offset:0}, compacted_segment=0, finished_self_compaction=0, finished_windowed_compaction=0, generation=3, reader={/var/lib/redpanda/data/kafka/topic-fadnsotldv/6_28/0-1-v1.log, (155 bytes)}, writer={closed:true, fallocation_offset:155, stable_offset:155, flushed_offset:155, committed_offset:155, inflight: 0, bytes_flush_pending:0, stats: { merged_writes: 0, bytes_requested : 155, bytes_written : 4096, appends : 2, flushes: 2, fsyncs: 3, truncates: 1, fallocations: 1, last_page_hydrations: 0}}, cache={cache_size=0, dirty tracker: {min: -9223372036854775808, max: -9223372036854775808}}, compaction_index:nullopt, closed=1, tombstone=0, index={file:/var/lib/redpanda/data/kafka/topic-fadnsotldv/6_28/0-1-v1.base_index, offsets:0, index:{header_bitflags:0, base_offset:0, max_offset:0, base_timestamp:{timestamp: 1762187916005}, max_timestamp:{timestamp: 1762187916005}, batch_timestamps_are_monotonic:1, with_offset:true, non_data_timestamps:1, broker_timestamp:{nullopt}, num_compactible_records_appended:{0}, clean_compact_timestamp:{nullopt}, may_have_tombstone_records:1, self_compact_timestamp:{nullopt}, may_have_transaction_batches:1, index(1, 1, 1)}, step:32768, needs_persistence:0}}]}], config: {ntp: {kafka/topic-fadnsotldv/6}, base_dir: /var/lib/redpanda/data, overrides: {compaction_strategy: {nullopt}, cleanup_policy_bitflags: {delete}, segment_size: {nullopt}, retention_bytes: {{nullopt}}, retention_time_ms: {{nullopt}}, recovery_enabled: false, retention_local_target_bytes: {{nullopt}}, retention_local_target_ms: {{nullopt}}, remote_delete: {true}, segment_ms: {{nullopt}}, initial_retention_local_target_bytes: {{nullopt}}, initial_retention_local_target_ms: {{nullopt}}, write_caching: {nullopt}, flush_ms: {nullopt}, flush_bytes: {nullopt}, iceberg_mode: disabled, remote_allow_gaps: {nullopt}, delete_retention_ms: {disabled}, min_cleanable_dirty_ratio: {{nullopt}}, min_compaction_lag_ms: {nullopt}, max_compaction_lag_ms: {nullopt} }}, revision: 28, topic_revision: 28, remote_revision: 28}}
```

WillemKauf reviewed (dismissed):
Is there a point to just going all the way on the `vassert() -> throw_if_closed()` conversion for all possible APIs that might be accessed by `consensus`?

andrwng committed:
storage: replace a few vasserts with throws

Recent changes added throw_if_closed() to address races between Kafka
layer callers and log shutdown, which is particularly important because
partitions may shutdown while Kafka layer calls are in flight.

That said, there are other calls into the log that originate from
outside the storage layer, and it isn't obvious that we never race with
a closing log there either. We hope there aren't, because the other
callers are typically Raft or cluster::partition, and we shut down the
Raft instance and partition before shutting down the log. But this
commit is an attempt to cover our bases just in case, throwing in other
calls that stem from these other non-Kafka layers.

Those are the following, which are all called from Raft.
- append
- truncate
- prefix truncate

andrwng commented:
>Is there a point to just going all the way on the vassert() -> throw_if_closed() conversion for all possible APIs that might be accessed by consensus?

Not a bad suggestion, given it's almost certainly going to be better than crashing.

I'd focused on the ones that I did because they originate in the Kafka layer, outside of {partition, log, raft}, whereas AFAICT everything else comes from within these subsystems. We shut these subsystems down in order of raft->partition->log, where, since the log is shutdown last, we wouldn't expect generally calls into the log originating from `raft` or `partition` to ever be performed on a closed log.

In any case, I'll make the change for calls that appear to not originate from the storage layer (append, truncate, prefix_truncate)

WillemKauf reviewed (approved):
Safety :safety_vest:

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5ee8d2d0-da4e-4e66-8905-8328225e4258-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76264">build#76264</a></summary>


| test_class                          | test_method                                  | test_arguments                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestSchemaRegistryACLs      | test_sr_audit_authz                          | {"audit_transport_mode": "kclient", "endpoint_name": "PUT_CONFIG_SUBJECT"} | integration | https://buildkite.com/redpanda/redpanda/builds/76264#019a7f36-f4b9-45ef-90f3-ce24c58797bf | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestSchemaRegistryACLs&test_method=test_sr_audit_authz         |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/76264#019a7f37-e216-4b9f-a1f6-bf6fd3fca260 | FLAKY         | 19/21    | upstream reliability is '92.61404779145546'. current run reliability is '90.47619047619048'. drift is 2.13786 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}                                                | integration | https://buildkite.com/redpanda/redpanda/builds/76264#019a7f37-e217-40c4-be8f-a4adfa2d535d | FLAKY         | 20/21    | upstream reliability is '91.0295616717635'. current run reliability is '95.23809523809523'. drift is -4.20853 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all         |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28528-v25.2.x-953 remotes/upstream/v25.2.x
git cherry-pick -x ee648559cc 74b79782fe
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19349996753).

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28528-v25.1.x-888 remotes/upstream/v25.1.x
git cherry-pick -x ee648559cc 74b79782fe
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19349996762).
